# STA 141B: Homework 1
Winter 2018

## Information

After the colons (in the same line) please write just your first name, last name, and the 9 digit student ID number below.

First Name: Zhihao

Last Name: Meng

Student ID: 915537642



## Instructions

We use a script that extracts your answers by looking for cells in between the cells containing the exercise statements.  So you 

- MUST add cells in between the exercise statements and add answers within them and
- MUST NOT modify the existing cells, particularly not the problem statement

To make markdown, please switch the cell type to markdown (from code) - you can hit 'm' when you are in command mode - and use the markdown language.  For a brief tutorial see: https://daringfireball.net/projects/markdown/syntax


## Part 1: The Doomsday Algorithm

The Doomsday algorithm, devised by mathematician J. H. Conway, computes the day of the week any given date fell on. The algorithm is designed to be simple enough to memorize and use for mental calculation.

__Example.__ With the algorithm, we can compute that July 4, 1776 (the day the United States declared independence from Great Britain) was a Thursday.

The algorithm is based on the fact that for any year, several dates always fall on the same day of the week, called the <em style="color:#F00">doomsday</em> for the year. These dates include 4/4, 6/6, 8/8, 10/10, and 12/12.

__Example.__ The doomsday for 2016 is Monday, so in 2016 the dates above all fell on Mondays. The doomsday for 2017 is Tuesday, so in 2017 the dates above will all fall on Tuesdays.

The doomsday algorithm has three major steps:

1. Compute the anchor day for the target century.
2. Compute the doomsday for the target year based on the anchor day.
3. Determine the day of week for the target date by counting the number of days to the nearest doomsday.

Each step is explained in detail below.

### The Anchor Day

The doomsday for the first year in a century is called the <em style="color:#F00">anchor day</em> for that century. The anchor day is needed to compute the doomsday for any other year in that century. The anchor day for a century $c$ can be computed with the formula:
$$
a = \bigl( 5 (c \bmod 4) + 2 \bigr) \bmod 7
$$
The result $a$ corresponds to a day of the week, starting with $0$ for Sunday and ending with $6$ for Saturday.

__Note.__ The modulo operation $(x \bmod y)$ finds the remainder after dividing $x$ by $y$. For instance, $12 \bmod 3 = 0$ since the remainder after dividing $12$ by $3$ is $0$. Similarly, $11 \bmod 7 = 4$, since the remainder after dividing $11$ by $7$ is $4$.

__Example.__ Suppose the target year is 1954, so the century is $c = 19$. Plugging this into the formula gives
$$a = \bigl( 5 (19 \bmod 4) + 2 \bigr) \bmod 7 = \bigl( 5(3) + 2 \bigr) \bmod 7 = 3.$$
In other words, the anchor day for 1900-1999 is Wednesday, which is also the doomsday for 1900.

__Exercise 1.1.__ Write a function that accepts a year as input and computes the anchor day for that year's century. The modulo operator `%` and functions in the `math` module may be useful. Document your function with a docstring and test your function for a few different years.  Do this in a new cell below this one.

In [4]:
def acrday(n):
    c=n//100
    a=(5*(c%4)+2)%7
    return a

In [5]:
acrday(2018)

2

### The Doomsday

Once the anchor day is known, let $y$ be the last two digits of the target year. Then the doomsday for the target year can be computed with the formula:
$$d = \left(y + \left\lfloor\frac{y}{4}\right\rfloor + a\right) \bmod 7$$
The result $d$ corresponds to a day of the week.

__Note.__ The floor operation $\lfloor x \rfloor$ rounds $x$ down to the nearest integer. For instance, $\lfloor 3.1 \rfloor = 3$ and $\lfloor 3.8 \rfloor = 3$.

__Example.__ Again suppose the target year is 1954. Then the anchor day is $a = 3$, and $y = 54$, so the formula gives
$$
d = \left(54 + \left\lfloor\frac{54}{4}\right\rfloor + 3\right) \bmod 7 = (54 + 13 + 3) \bmod 7 = 0.
$$
Thus the doomsday for 1954 is Sunday.

__Exercise 1.2.__ Write a function that accepts a year as input and computes the doomsday for that year. Your function may need to call the function you wrote in exercise 1.1. Make sure to document and test your function.

In [6]:
def doomsday(m):
    a=acrday(m)
    y=m%100
    d=(y+int(y/4)+a)%7
    return d

In [7]:
doomsday(2018)

3

### The Day of Week

The final step in the Doomsday algorithm is to count the number of days between the target date and a nearby doomsday, modulo 7. This gives the day of the week.

Every month has at least one doomsday:
* (regular years) 1/10, 2/28
* (leap years) 1/11, 2/29
* 3/21, 4/4, 5/9, 6/6, 7/11, 8/8, 9/5, 10/10, 11/7, 12/12

__Example.__ Suppose we want to find the day of the week for 7/21/1954. The doomsday for 1954 is Sunday, and a nearby doomsday is 7/11. There are 10 days in July between 7/11 and 7/21. Since $10 \bmod 7 = 3$, the date 7/21/1954 falls 3 days after a Sunday, on a Wednesday.

__Exercise 1.3.__ Write a function to determine the day of the week for a given day, month, and year. Be careful of leap years! Your function should return a string such as "Thursday" rather than a number. As usual, document and test your code.

In [8]:
def weekday(date):
    a=date
    b=a.split('/')
    c=[int(x) for x in b]
    d1=[10,28,21,4,9,6,11,8,5,10,7,12]
    d2=[11,29,21,4,9,6,11,8,5,10,7,12]
    w=['Sunday','Monday','Tuesday','Wednesday','Thursday','Friday','Saturday']
    if c[2]%100==0:
        if c[2]%400==0:
            md=d2[c[0]-1]#Find the doomsday of target month
            dd=doomsday(c[2])#Find the number of weekday of the doomsday
            wd=((c[1]-md)%7+dd)%7#Find the number of weekday of target day
            d=w[wd]#Find the weekday of tearget day
        else:
            md=d1[c[0]-1]#Find the doomsday of target month
            dd=doomsday(c[2])#Find the number of weekday of the doomsday
            wd=((c[1]-md)%7+dd)%7#Find the number of weekday of target day
            d=w[wd]#Find the weekday of tearget day
    else:
        if c[2]%4==0:
            md=d2[c[0]-1]#Find the doomsday of target month
            dd=doomsday(c[2])#Find the number of weekday of the doomsday
            wd=((c[1]-md)%7+dd)%7#Find the number of weekday of target day
            d=w[wd]#Find the weekday of tearget day
        else:
            md=d1[c[0]-1]#Find the doomsday of target month
            dd=doomsday(c[2])#Find the number of weekday of the doomsday
            wd=((c[1]-md)%7+dd)%7#Find the number of weekday of target day
            d=w[wd]#Find the weekday of tearget day
    return d

In [9]:
weekday('5/8/2017')

'Monday'

__Exercise 1.4.__ How many times did Friday the 13th occur in the years 1900-1999? Does this number seem to be similar to other centuries?

In [10]:
def Friday(y1,y2):#Input the first year and the last year
    da=[]
    twd=[]
    for i in list(range(y1,y2+1)):#Creat a list includes the date: Month/13/Year
        for j in list(range(1,13)):
            x=str(j)+'/13/'+str(i)
            da.append(x)
    for k in list(range(len(da))):#Find the weekday of all the 13th between y1 and y2
        y=weekday(da[k])
        twd.append(y)
    friday=[z for z in twd if z=='Friday']
    num_friday=len(friday)#Find the number of Friday the 13th occur between y1 and y2
    return num_friday

In [11]:
Friday(1900,1999)

172

In [12]:
Friday(1900,1999)

172

In [13]:
Friday(800,899)

172

The number seems to be very similar with other centuries.

__Exercise 1.5.__ How many times did Friday the 13th occur between the year 2000 and today?

In [14]:
def FridayToday(date1):
    e=date1
    f=e.split('/')
    g=[int(x) for x in f]
    da1=[]
    twd1=[]
    for i in list(range(2000,g[2])):#Creat a list includes the date: Month/13/Year from 2000 to last year
        for j in list(range(1,13)):
            x=str(j)+'/13/'+str(i)
            da1.append(x)
    if g[1]>13:#Decide the latest date is 13th this month or last month according the fact if 13th this month have passed
        for k in list(range(1,g[0]+1)):
            x1=str(k)+'/13/'+str(g[2])
            da1.append(x1)
    else:
        for l in list(range(1,g[0])):
            x2=str(l)+'/13/'+str(g[2])
            da1.append(x2)
    for m in list(range(len(da1))):#Find the weekday of all the 13th during this period
        y=weekday(da1[m])
        twd1.append(y)
    friday1=[z for z in twd1 if z=='Friday']
    num_friday1=len(friday1)#Find the number of Friday the 13th occur during this period
    return num_friday1

In [15]:
FridayToday('1/19/2018')

31

## Part 2: 1978 Birthdays

__Exercise 2.1.__ The file `birthdays.txt` contains the number of births in the United States for each day in 1978. Inspect the file to determine the format. Note that columns are separated by the tab character, which can be entered in Python as `\t`. Write a function that uses iterators and list comprehensions with the string methods `split()` and `strip()` to  convert each line of data to the list format

```Python
[month, day, year, count]
```
The elements of this list should be integers, not strings. The function `read_birthdays` provided below will help you load the file.

In [1]:
def read_birthdays(file_path):
    """Read the contents of the birthdays file into a string.
    
    Arguments:
        file_path (string): The path to the birthdays file.
        
    Returns:
        string: The contents of the birthdays file.
    """
    with open(file_path) as file:
        return file.read()

In [16]:
def read_birthdays(file_path):
    with open(file_path) as file:
        return file.read()

In [17]:
def birthdays(file_path):
    data=read_birthdays(file_path)
    data1=str(data)
    data2=data1.split('\n')#Split the data by \n
    del data2[:6]#Delete the first 6 lines of the data
    for i in range(data2.count('')):#Delete all ' ' in the data
        data2.pop()
    for j in range(data2.count('\t')):#Delete all '\t' in the data
        data2.pop()
    data3=[x.split() for x in data2]#Split every line by \t
    for k in range(len(data3)):#Split the Month/Day/Year by '/'
        y=data3[k][0].split('/')
        y.append(data3[k][1])
        data3[k]=[int(z) for z in y]#Transfor the data from string to integer
    return data3

In [18]:
data3=birthdays('E:/ucdavis/lecture/STA141B/hw1/birthdays.txt')
data3

[[1, 1, 78, 7701],
 [1, 2, 78, 7527],
 [1, 3, 78, 8825],
 [1, 4, 78, 8859],
 [1, 5, 78, 9043],
 [1, 6, 78, 9208],
 [1, 7, 78, 8084],
 [1, 8, 78, 7611],
 [1, 9, 78, 9172],
 [1, 10, 78, 9089],
 [1, 11, 78, 9210],
 [1, 12, 78, 9259],
 [1, 13, 78, 9138],
 [1, 14, 78, 8299],
 [1, 15, 78, 7771],
 [1, 16, 78, 9458],
 [1, 17, 78, 9339],
 [1, 18, 78, 9120],
 [1, 19, 78, 9226],
 [1, 20, 78, 9305],
 [1, 21, 78, 7954],
 [1, 22, 78, 7560],
 [1, 23, 78, 9252],
 [1, 24, 78, 9416],
 [1, 25, 78, 9090],
 [1, 26, 78, 9387],
 [1, 27, 78, 8983],
 [1, 28, 78, 7946],
 [1, 29, 78, 7527],
 [1, 30, 78, 9184],
 [1, 31, 78, 9152],
 [2, 1, 78, 9159],
 [2, 2, 78, 9218],
 [2, 3, 78, 9167],
 [2, 4, 78, 8065],
 [2, 5, 78, 7804],
 [2, 6, 78, 9225],
 [2, 7, 78, 9328],
 [2, 8, 78, 9139],
 [2, 9, 78, 9247],
 [2, 10, 78, 9527],
 [2, 11, 78, 8144],
 [2, 12, 78, 7950],
 [2, 13, 78, 8966],
 [2, 14, 78, 9859],
 [2, 15, 78, 9285],
 [2, 16, 78, 9103],
 [2, 17, 78, 9238],
 [2, 18, 78, 8167],
 [2, 19, 78, 7695],
 [2, 20, 78, 9021]

__Exercise 2.2.__ Which month had the most births in 1978? Which day of the week had the most births? Which day of the week had the fewest? What conclusions can you draw? You may find the `Counter` class in the `collections` module useful.

In [19]:
all_month=[]
all_count=[]
for i in range(1,13):#Caculate the sum of birth for every month
    month=[]
    month_count=[]
    for j in range(len(data3)):
        if i==data3[j][0]:
            month.append(data3[j][3])#Record the birth for every month
    all_count.append(sum(month))
    month_count.append(sum(month))
    month_count.append(i)
    all_month.append(month_count)
print(all_month)
for k in range(len(all_month)):#Find the month with most birth
    if all_month[k][0]==max(all_count):
        print(all_month[k])

[[270695, 1], [249875, 2], [276584, 3], [254577, 4], [270812, 5], [270756, 6], [294701, 7], [302795, 8], [293891, 9], [288955, 10], [274671, 11], [284927, 12]]
[302795, 8]


In [21]:
date1=[]
day=[]
day_num=[]
weekday_num=[]
all_weekday=[]
for i in range(len(data3)):#Change the form of date to Month/Day/1978   
    a=str(data3[i][0])+'/'+str(data3[i][1])+'/'+'19'+str(data3[i][2])
    date1.append(a)
for j in range(len(date1)):#Find the weekday for every day in 1978  
    b=weekday(date1[j])
    day.append(b)
for k in range(len(day)):#Match the birth with weekday
    c=[day[k],data3[k][3]]
    day_num.append(c)
week=['Sunday','Monday','Tuesday','Wednesday','Thursday','Friday','Saturday']
for l in range(len(week)):#Calculate the sum of birth for every weekday
    week1=[]
    weekday1=[]
    for m in range(len(day_num)):#Record the birth for every week day
        if day_num[m][0]==week[l]:
            week1.append(day_num[m][1])
    weekday_num.append(sum(week1))
    weekday1.append(week[l])
    weekday1.append(sum(week1))
    all_weekday.append(weekday1)
print(all_weekday)
max_weekday=[]
for n in range(len(weekday_num)):#Find the weekday with most birth
    if weekday_num[n]==max(weekday_num):
        max_weekday.append(week[n])
        max_weekday.append(max(weekday_num))
print(max_weekday)  
min_weekday=[]
for p in range(len(weekday_num)):#Find the week day with fewest birth
    if weekday_num[p]==min(weekday_num):
        min_weekday.append(week[p])
        min_weekday.append(min(weekday_num))
print(min_weekday)

[['Sunday', 421400], ['Monday', 487309], ['Tuesday', 504858], ['Wednesday', 493897], ['Thursday', 493149], ['Friday', 500541], ['Saturday', 432085]]
['Tuesday', 504858]
['Sunday', 421400]


The month which has most birth occur is August and the total number is 302795

The weekday which has most birth occur is Tuesday and the total number is 504858. The weekday which has fewest is Sunday and the total number is 421400.

__Exercise 2.3.__ What would be an effective way to present the information in exercise 2.2? You don't need to write any code for this exercise, just discuss what you would do.

I think a table or chart which including the weekday and the number of birth on that weekday is a more effective way to present the information in exercise 2.2. For example, a line chart can show the change of number of birth for every month and we can find the maximum and minimum directly.